plutus-core/plutus-ir/src/PlutusIR/Parser.hs

```diff
@@ -29,109 +28,113 @@ import PlutusCore.Error (AsParserErrorBundle)
 import Text.Megaparsec hiding (ParseError, State, many, parse, some)
 
 -- | A parsable PIR pTerm.
-type PTerm = PIR.Term TyName Name PLC.DefaultUni PLC.DefaultFun SourcePos
+type PTerm = PIR.Term TyName Name PLC.DefaultUni PLC.DefaultFun SrcSpan
 
 recursivity :: Parser Recursivity
-recursivity = inParens $ (wordPos "rec" >> return Rec) <|> (wordPos "nonrec" >> return NonRec)
+recursivity = inParensSpc $ (symbol "rec" $> Rec) <|> (symbol "nonrec" $> NonRec)
```

Comment:
I think we need a note about whitespace handling. Initially I thought all uses of `inParensSpc` must be wrong, because that meant that either a) the whitespace would get included in the span or b) we weren't including the parens in the span, which we should be. Now I see that it can work, but only if we are parsing something that doesn't have a span annotation.

TBH, I think I'd find it clearer if we just used an explicit `*> whitespace` instead. Or a careful note on `inParensSpc`.

Generally I think it's not totally clear what you're supposed to do about whitespace. I think the answer is "all parsers should consume trailing whitespace, but it should not be included in the source span". `withSpan` does this automatically, which is good. Perhaps we should just use `withSpan` for everything, and just ignore the span if we don't have a place for it?

Reply:
> I think I'd find it clearer if we just used an explicit *> whitespace

I'm fine with that, but like I said above, I don't think it reduces the chance of doing the wrong thing. One could easily forget to add `*> whitespace`.

> Perhaps we should just use withSpan for everything, and just ignore the span if we don't have a place for it?

This would make it cleaner, but it goes against the recommendation, which is that `getSourcePos` is expensive, and should not be used more than needed.

The cleanest solution is to use a lexer like you've suggested before.

Reply:
I added helper functions `leadingWhitespace` and `trailingWhitespace`, and replaced `inParensSpc` with `trailingWhitespace . inParens`.